### ISSUE TITLE
Failing Automations: Homebridge Restart Removes Listeners for Characteristic Events and HomeKit Hub Never Re-Registers

### ISSUE THREAD

**Issue by nikonratm** [nfarina/homebridge#2413]
### Pre Steps
Also raised this at: https://github.com/KhaosT/HAP-NodeJS/issues/789 but no response there...

I've discussed this with a fellow Homebridge plugin dev and we are both reproducing this issue on tvOS 13.3.1, see: https://github.com/nzapponi/homebridge-simplisafe3/issues/66#issuecomment-581301311

In short, after restarting Homebridge (e.g. just restarting the service) a HomeKit hub (Apple TV 4k) never "re-registers" as a listener to characteristic events, not until it is manually restarted. Sometimes even a restart isnt enough, I need to manually toggle Settings > Airplay & HomeKit and toggle my home from Connected / Disconnected and then restart. Huh?

Is this a purely tvOS issue? Known? I'm not even 100% clear about which side of things initiates registering a device as a listener—is it totally reliant that the HomeKit device handles this in the same way that launching / quitting Home.app on iOS devices takes care of it? Should HAP-NodeJS should be "reconstructing" the list of listeners after it restarts? Or finally, is there some issue about Homebridge / HAP-NodeJS 'announcing' its shutdown/restart in some way that would 'remind' the HomeKit hub to re-subscribe to events?

Appreciate any thoughts / insight. If there is anything that can be done on this side of things would love to help.

### Expected Behavior
HomeKit hub should be re-registered for events and continue receiving events after Homebridge restart.

### Current Behavior
HomeKit hub (Apple TV) never re-reigsters for events and all automations related to Homebridge start failing. Sometimes restart corrects or sometimes hub needs to be 'disconnected' from Home in Apple TV settings.

### Steps to Reproduce
See https://github.com/nzapponi/homebridge-simplisafe3/issues/66#issuecomment-581301311
1. Set up an automation based on a contact sensor
2. Close Home on all devices and disconnect from Wi-Fi
3. Restart Homebridge: **Apple TV won't register for characteristic events**
4. Open the contact sensor - the automation will not work.
5. Restart Apple TV - when it comes back on, usually it does register for characteristic events (but not always!)
6. Open the contact sensor - automation should work

### Versions
- **Homebridge**: 0.4.50
- **Node**:       13.7.0
- **OS**:         Raspbian GNU/Linux 10 (buster)
- **HomeKit Hub**:         ATV 4K on tvOS 13.3.1

### Homebridge Log / Command Output
```
Fri, 31 Jan 2020 08:47:45 GMT EventedHTTPServer [Apple TV Static IP] HTTP request: /characteristics
Fri, 31 Jan 2020 08:47:45 GMT HAPServer [xx:xx:xx:xx:xx:xx] HAP Request: PUT /characteristics
Fri, 31 Jan 2020 08:47:45 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":8,"iid":9,"ev":true},{"aid":8,"iid":10,"ev":true}]
Fri, 31 Jan 2020 08:47:45 GMT Accessory [Homebridge] Registering Characteristic "Contact Sensor State" for events
Fri, 31 Jan 2020 08:47:45 GMT Accessory [Homebridge] Registering Characteristic "Status Low Battery" for events
```

**Comment by nikonratm** [comment]
Just makes no sense 😖 I just restarted my ATV and toggled 'connected/disconnected' under Airplay & HomeKit settings and it _still_ didn't re-register. After I restarted it I see a few of my iOS devices register, but not the ATV.

**Comment by HunterDG** [comment]
Wonder if this is related to HomePod failing to function as a remote gateway when it takes over from AppleTV

**Comment by nikonratm** [comment]
I don’t have a HomePod! Just ATV :)

**Comment by HunterDG** [comment]
behavior seems too similar to be coincidence

**Comment by nikonratm** [comment]
Similar to what? Are there similar reports for people with HomePod (link?)? I know technically they run different OS but I would hope their HomeKit hub code is similar so yea I could def see this being a problem with HomePod as well, I was just pointing out that I only have 1 hub on my network not multiple.

**Comment by ebaauw** [comment]
Does the AppleTV fail to connect to homebridge at all, or does it connect but not register for changing characteristic values?  Can you check with `netstat` on the homebridge server before and after restarting homebridge?

If the AppleTV does connect, I would suspect a bug in the AppleTV.  If it doesn’t connect, I would suspect network issues with the Bonjour announcements by homebridge not reaching the AppleTV.  How is your AppleTV connected to the network?  How is homebridge?  What routers and switches do you have in between them?  Do they forward multicast traffic unconditionally to all ports, or are they trying to be smart (IGMP snooping) and only forward it to the ports with a listener?

**Comment by nikonratm** [comment]
Thanks for the reply. After I posted this I tried turning off wifi on my RPi completely which seemed to make a difference, ATV now seems to be better about eventually reconnecting at least while Homebridge is not restarted. With WiFi on I noticed that the RPi was broadcasting HAP on Bonjour with two IP addresses (elan and wlan, both of which were set to static IP) now its just the one.

BUT, theres still an issue where ATV never reconnects after HB restart and in general it takes a long time to reconnect if it ever does eventually. Trying what you suggested @ebaauw , before I restart HB `netstat` shows:
`tcp        0      0 x.x.x.130:33209        x.x.x.120:49394        ESTABLISHED`
whereas afterwards it doesnt list any connection to `x.x.x.120` which is ATV IP. But when HB restarts I immediately see /pair-verify requests coming from other iOS devices on my network—so why are those 'aware' of the bridge restart and better about reconnecting than the ATV??

The other thing that happened is that last night my 1AM automation failed because the ATV 'disconnected' just before 1AM, and didn't connect until after. Disconnects seem to happen often now, even with WiFi turned off. Why would they? And why does it take so long to reconnect. E.g. just before 1AM I see:
```
2/7/2020, 12:52:42 AM [x.x.x.120] HTTP connection was closed
2/7/2020, 12:52:42 AM [x.x.x.120] HTTP connection was closed
2/7/2020, 12:53:28 AM [x.x.x.120] HTTP connection was closed
```
Then _after_ 1AM the ATV 'comes back', but of course its too late for the 1AM automation:
```
2/7/2020, 1:12:52 AM [x.x.x.120] HTTP request: /pair-verify
2/7/2020, 1:12:52 AM [x.x.x.120] HTTP request: /pair-verify
2/7/2020, 1:12:53 AM [x.x.x.120] HTTP request: /pair-verify
2/7/2020, 1:12:53 AM [x.x.x.120] HTTP request: /pair-verify
2/7/2020, 1:12:53 AM [x.x.x.120] HTTP request: /pair-verify
2/7/2020, 1:12:53 AM [x.x.x.120] HTTP request: /pair-verify
```
In between those two I see a series of `keepalive` requests to iOS devices, maybe they are not happening frequently enough to keep the ATV connection going?

Basically seems that ATV sucks at reconnecting, whether its from random disconnects or from Homebridge restart. My hope is how to mitigate this if its possible from the Homebridge side but perhaps it is just down to the ATV software itself?? Frustrating...

Both ATV and RPi are hardwired ethernet. There is 1 ethernet switch between them. All devices are behind an Airport Extreme router. I have IPV6 disabled on the RPi. IGMP snooping is disabled on the router.

Really appreciate everyone's input.

**Comment by ebaauw** [comment]
> All devices are behind an Airport Extreme router.

You mean your iOS devices?  What happens if you connect the AppleTV the same way (wirelessly instead of wired)?

> There is 1 ethernet switch between them.

Is that a dumb switch or a managed one?  In the latter case: did you turn off IGMP snooping on the switch as well?  Is the Airport Extreme also connected to this switch?

**Comment by nikonratm** [comment]
For example I just restarted HB and ATV didn't disconnect so I went into the ATV, restarted it and toggled my 'Home' connected / disconnected and it _still_ didn't reconnect. Every time I 'disconnected' and 'connected' the Apple TV (under settings) I just saw over and over my computer and iPad and iPhone IP addresses 'registering' for characteristic events—never the ATV. Then I went and turned off Wifi on my iPad and then suddenly the ATV registers.

And now I literally cannot get the ATV to register. I turned off all other devices, restarted the ATV and I still dont see it register, when I reconnect my laptop to WiFi and look at the logs it shows my laptop registering as a listener, nothing from the ATV. It's just crazy!?

All switches are unmanaged.

**Comment by llamafilm** [comment]
This is happening to me, even without restarting homebridge.  I have a HomePod running iOS 13.3.1 and a StatelessProgrammableSwitch provided by [homebridge-lutron-caseta](https://github.com/smockle/homebridge-lutron-caseta) plugin.  When I first launch homebridge, it never registers the buttons until I open Home.app (or control center).  Then about 2 minutes later, it unregisters the buttons.

Let me know if I should open separate issue for this but it seems related.

```
Sat, 08 Feb 2020 21:02:04 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:02:04 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":2,"iid":10,"ev":true},{"aid":2,"iid":13,"ev":true},{"aid":2,"iid":16,"ev":true},{"aid":2,"iid":19,"ev":true},{"aid":2,"iid":22,"ev":true}]
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
......
Sat, 08 Feb 2020 21:03:30 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:03:30 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":2,"iid":10,"ev":false},{"aid":2,"iid":13,"ev":false},{"aid":2,"iid":16,"ev":false},{"aid":2,"iid":19,"ev":false},{"aid":2,"iid":22,"ev":false}]
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:03:31 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":1,"iid":9,"ev":false},{"aid":1,"iid":10,"ev":false},{"aid":1,"iid":11,"ev":false}]
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "State" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Version" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Control Point" for events
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
```

**Comment by nikonratm** [comment]
I'm almost embarrassed to report but I think this had a lot to do with my network setup. Specifically I had hub → switch → switch (RPi with Homebridge off this switch) → router. I changed it so all switches are direct to router e.g. the ATV has only 1 switch between it and router (though I will point still only 1 switch between it and the RPi). Anyway, this seems to have drastically improved the reliability. I now see a tiny fraction of the number of 'disconnects' from my ATV and it even seems to re-connect faster after HB restart.

I'll admit I'm a bit frustrated its something I hadn't even really considered but thanks to @ebaauw for putting me on the right path! I think this can be closed now (and I can finally move on with my life 🤪)

**Comment by nikonratm** [comment]
@llamafilm can you identify for us the IP addresses in your logs? Is one of those your HomePod?

**Comment by llamafilm** [comment]
IP .107 is the iPhone where I opened Home.app
IP .180 is the HomePod
IP .252 is another iPad, not in use right now

I didn't think the lines in between were relevant but here they are just in case.  All of these devices are on WiFi connected to the same router.

```
Sat, 08 Feb 2020 21:02:04 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:02:04 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":2,"iid":10,"ev":true},{"aid":2,"iid":13,"ev":true},{"aid":2,"iid":16,"ev":true},{"aid":2,"iid":19,"ev":true},{"aid":2,"iid":22,"ev":true}]
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT Accessory [Homebridge] Registering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:02:04 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
Sat, 08 Feb 2020 21:02:15 GMT EventedHTTPServer [::ffff:192.168.3.180] New connection from client
Sat, 08 Feb 2020 21:02:15 GMT EventedHTTPServer [::ffff:192.168.3.180] HTTP server listening on port 35849
Sat, 08 Feb 2020 21:02:15 GMT EventedHTTPServer [::ffff:192.168.3.180] HTTP request: /pair-verify
Sat, 08 Feb 2020 21:02:15 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: POST /pair-verify
Sat, 08 Feb 2020 21:02:15 GMT HAPServer [CB:22:3D:E2:CE:31] Pair verify step 1/2
Sat, 08 Feb 2020 21:02:16 GMT EventedHTTPServer [::ffff:192.168.3.180] HTTP Response is finished
Sat, 08 Feb 2020 21:02:16 GMT EventedHTTPServer [::ffff:192.168.3.180] Client connection closed
Sat, 08 Feb 2020 21:02:16 GMT EventedHTTPServer [::ffff:192.168.3.180] HTTP connection was closed
Sat, 08 Feb 2020 21:02:16 GMT EventedHTTPServer [::ffff:192.168.3.180] HTTP server was closed
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] New connection from client
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP server listening on port 39217
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP request: /pair-verify
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: POST /pair-verify
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] Pair verify step 1/2
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP Response is finished
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP request: /pair-verify
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: POST /pair-verify
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] Pair verify step 2/2
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] Client F709FBA4-2080-4351-9C50-6A271B0DAD50 verification complete
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP Response is finished
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP request: /accessories
Sat, 08 Feb 2020 21:02:50 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: GET /accessories
Sat, 08 Feb 2020 21:02:50 GMT EventedHTTPServer [::ffff:192.168.3.252] HTTP Response is finished
Sat, 08 Feb 2020 21:03:30 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:03:30 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":2,"iid":10,"ev":false},{"aid":2,"iid":13,"ev":false},{"aid":2,"iid":16,"ev":false},{"aid":2,"iid":19,"ev":false},{"aid":2,"iid":22,"ev":false}]
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:30 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Programmable Switch Event" for events
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP request: /characteristics
Sat, 08 Feb 2020 21:03:31 GMT HAPServer [CB:22:3D:E2:CE:31] HAP Request: PUT /characteristics
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Processing characteristic set: [{"aid":1,"iid":9,"ev":false},{"aid":1,"iid":10,"ev":false},{"aid":1,"iid":11,"ev":false}]
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "State" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Version" for events
Sat, 08 Feb 2020 21:03:31 GMT Accessory [Homebridge] Unregistering Characteristic "Control Point" for events
Sat, 08 Feb 2020 21:03:31 GMT EventedHTTPServer [::ffff:192.168.3.107] HTTP Response is finished
```

**Comment by nikonratm** [comment]
Hmm, I think it is normal for iOS devices to un-register periodically (note that .107 is the thing in your logs unregistering, not .180), your home hub should be the one to never unregister (and in your log excerpts, it doesnt look like it is). I dont have experience with HomePod as a hub (I'm assuming its your hub) as opposed to ATV but my assumption is they work the same. 

Do you have any other Homebridge plugins? For example even https://www.npmjs.com/package/homebridge-dummy . Is it possible its an issue with the lutron plugin?

**Comment by llamafilm** [comment]
It certainly could be a problem with the lutron plugin.  Unfortunately the author has abandoned it and I'm not very good with JS.  This dummy plugin looks promising.  I'll test that and if it works maybe I can figure out how to fix the lutron code from there.
